lib/px/service/kinesis/base_request.rb

```diff
@@ -33,15 +33,67 @@ def initialize
       @last_send = Time.now
       @last_throughput_exceeded = nil
 
-      @buffer = Array.new
+      @buffer = []
     end
 
     ##
     # Check if buffer should be flushed and sent to kinesis
-    #
     def flush_records
-      if (put_rate_decay == DEFAULT_PUT_RATE && @buffer.length >= FLUSH_LENGTH) || (Time.now - @last_send > put_rate_decay)
-        response = @kinesis.put_records(:stream_name => @stream, :records => @buffer)
+      begin
+        _flush_records
+      rescue Px::Service::ServiceError, Px::Service::ServiceRequestError => ex
+        raise ex
+      rescue StandardError => ex
+        # Wrap other exceptions, includes CircuitBreaker::CircuitBrokenException
+        raise Px::Service::ServiceError.new(ex.message, 503), ex, ex.backtrace
+      end
+    end
+
+    ##
+    # Takes a blob of data to send to Kinesis.
+    # The data will be msgpacked and queued for send.
+    #
+    # Returns the number of unsent messages
+    def queue_record(data)
+      data_blob = data.to_msgpack
+
+      # TODO: ensure partition key is distributed over shards
+      @buffer << { data: data_blob, partition_key: Px::Service::Kinesis.config.partition_key }
+
+      # check if we should flush the buffer
+      flush_records
+
+      return @buffer.size
+    end
+
+    # push a single record to kinesis, bypass the buffer
+    #
+    # Please don't use unless necessary.
+    # Mainly used to bypass buffering when testing
+    def put_record(data)
+      begin
+        _put_record(data)
+      rescue Px::Service::ServiceError, Px::Service::ServiceRequestError => ex
+        raise ex
+      rescue StandardError => ex
+        # Wrap other exceptions, includes CircuitBreaker::CircuitBrokenException
+        raise Px::Service::ServiceError.new(ex.message, 503), ex, ex.backtrace
+      end
+    end
+
+    ##
+    # Returns true if the buffered messages can be flushed
+    def can_flush?
+      (put_rate_decay == DEFAULT_PUT_RATE && @buffer.size >= FLUSH_LENGTH) || (Time.now - @last_send > put_rate_decay)
+    end
+
+    private
+
+    ##
+    # Check if buffer should be flushed and sent to kinesis
+    def _flush_records
+      if can_flush?
+        response = @kinesis.put_records(stream_name: @stream, records: @buffer)
```

Comment:
If this returns any 400 (other than flow control) you should raise a `ServiceRequestError` exception.  If it returns any 500 (other than flow control), or times out, you should raise a `ServiceError`.  The circuit breaker depends on this distinction to operate correctly.